packages/aws-cdk/test/integ/cli/bootstrapping.integtest.ts

```diff
@@ -1,52 +1,175 @@
+const QUALIFIER = randomString();
 
 beforeAll(async () => {
   await prepareAppFixture();
 });
 
 beforeEach(async () => {
   await cleanupOldStacks();
+  BUCKETS_TO_DELETE.splice(0, BUCKETS_TO_DELETE.length);
 });
 
 afterEach(async () => {
   await cleanupOldStacks();
+  await cleanupBootstrapStacks();
+
+  // We might have lost some buckets by upgrading the bootstrap stack. Be
+  // sure to clean everything.
+  for (const bucket of BUCKETS_TO_DELETE) {
+    await deleteBucket(bucket);
+  }
 });
 
 test('can bootstrap without execution', async () => {
-  const bootstrapStackName = fullStackName('toolkit-stack-1');
+  const bootstrapStackName = `${BOOTSTRAP_STACK_PREFIX}-1`;
 
-  await cdk(['bootstrap', '-v',
+  await cdk(['bootstrap',
     '--toolkit-stack-name', bootstrapStackName,
     '--no-execute']);
+
+  const resp = await cloudFormation('describeStacks', {
+    StackName: bootstrapStackName,
+  });
+
+  expect(resp.Stacks?.[0].StackStatus).toEqual('REVIEW_IN_PROGRESS');
+});
+
+test('upgrade legacy bootstrap stack to new bootstrap stack while in use', async () => {
+  const bootstrapStackName = `${BOOTSTRAP_STACK_PREFIX}-update-in-place`;
+
+  const legacyBootstrapBucketName = `aws-cdk-bootstrap-integ-test-legacy-bckt-${randomString()}`;
+  const newBootstrapBucketName = `aws-cdk-bootstrap-integ-test-v2-bckt-${randomString()}`;
+  BUCKETS_TO_DELETE.push(legacyBootstrapBucketName, newBootstrapBucketName);
+
+  // Legacy bootstrap
+  await cdk(['bootstrap',
+    '--toolkit-stack-name', bootstrapStackName,
+    '--bootstrap-bucket-name', legacyBootstrapBucketName]);
+
+  // Deploy stack that uses file assets
+  await cdkDeploy('lambda', {
+    options: ['--toolkit-stack-name', bootstrapStackName],
+  });
+
+  // Upgrade bootstrap stack to "new" style
+  await cdk(['bootstrap',
+    '--toolkit-stack-name', bootstrapStackName,
+    '--bootstrap-bucket-name', newBootstrapBucketName,
+    '--qualifier', QUALIFIER], {
+    modEnv: {
+      CDK_NEW_BOOTSTRAP: '1',
+    },
+  });
+
+  // (Force) deploy stack again
+  // --force to bypass the check which says that the template hasn't changed.
+  await cdkDeploy('lambda', {
+    options: [
+      '--toolkit-stack-name', bootstrapStackName,
+      '--force',
+    ],
+  });
+});
+
+test('deploy new style synthesis to new style bootstrap', async () => {
+  const bootstrapStackName = BOOTSTRAP_STACK_PREFIX;
+
+  await cdk(['bootstrap',
+    '--toolkit-stack-name', bootstrapStackName,
+    '--qualifier', QUALIFIER], {
+    modEnv: {
+      CDK_NEW_BOOTSTRAP: '1',
+    },
+  });
+
+  // Deploy stack that uses file assets
+  await cdkDeploy('lambda', {
+    options: [
+      '--toolkit-stack-name', bootstrapStackName,
+      '--context', `@aws-cdk/core:bootstrapQualifier=${QUALIFIER}`,
+      '--context', '@aws-cdk/core:newStyleStackSynthesis=1',
+    ],
+  });
+});
+
+test('can bootstrap multiple legacy toolkit stacks', async () => {
+  const bootstrapStackName1 = `${BOOTSTRAP_STACK_PREFIX}-1`;
```

Comment:
I'm a little confused by these bootstrap stack names. At first I thought every test had their own, unique one, but `${BOOTSTRAP_STACK_PREFIX}-1` is re-used between 2 of them... if they don't have to be unique, can't we just use `BOOTSTRAP_STACK_PREFIX` instead?